api/types/maintenancewindow.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gravitational/trace"
+)
+
+const (
+	UpgraderKindKubeController = "kube"
+	UpgraderKindSystemdUnit    = "unit"
+)
+
+var validWeekdays = [7]time.Weekday{
+	time.Sunday,
+	time.Monday,
+	time.Tuesday,
+	time.Wednesday,
+	time.Thursday,
+	time.Friday,
+	time.Saturday,
+}
+
+// parseWeekday attempts to interpret a string as a time.Weekday. In the interest of flexibility,
+// parsing is case-insensitive and supports the common three-letter shorthand accepted by many
+// common scheduling utilites (e.g. contab, systemd timers).
+func parseWeekday(s string) (day time.Weekday, ok bool) {
+	for _, w := range validWeekdays {
+		if strings.EqualFold(w.String(), s) || strings.EqualFold(w.String()[:3], s) {
+			return w, true
+		}
+	}
+
+	return time.Sunday, false
+}
+
+// generator builds a closure that iterates valid maintenance window from the current day onward. Used in
+// schedule export logic and tests.
+func (w *AgentUpgradeWindow) generator(from time.Time) func() (start time.Time, end time.Time) {
+	from = from.UTC()
+	next := time.Date(
+		from.Year(),
+		from.Month(),
+		from.Day(),
+		int(w.UTCStartHour%24),
```

Comment:
Why we need mod 24 ? how the starting hours can be larger than 23 ?

Reply:
It won't be larger than 23 unless there is a bug in `CheckAndSetDefaults`.  This is just for the purposes of limiting the impact of such a bug by ensuring that we won't arbitrarily jump around/skip days if an invalid hour value does make it in somehow.  go's time package will just move the date arbitrarily forward to accommodate a larger than 23 hour value.  A bug in how we handle `UTCStartHour` resulting in the wrong hour being used in the timestamp is easy to debug... a bug resulting in wacky timestamps way off from the expected day is harder to debug.

Basically, I just do this cuz I dislike how go's time package handles bad inputs.